### ISSUE TITLE
How to calculate parameters and FLOPs?

### ISSUE THREAD

**Issue by zouyanmei** [Matrix-ASC/DeLA#9]
Thanks for your share. Could you share the code for calculating parameters, FLOPs and Throughout? 
Moreover, do you plan to extend to 6-fold on S3DIS, and how to test the DeLA without voting? (I got the 87.3 cls. miou and 85.8 cat mIoU, which is different to public results. )
Thanks again.

**Comment by binjiechen** [comment]
Hi.
① I mostly followed [PointNeXt#profile.py](https://github.com/guochengqian/PointNeXt/blob/master/examples/profile.py) for profiling. The basic logic is simple. Load a batch of inputs (I always load from dataloader), use deepspeed for FLOPs and params; do multiple forwarding to get throughput.
② Currently I'm occupied with other commitments. So at least for the time being, I have no plans for that (or even update the current version of code).
③ I assume you mean ShapeNetPart. The for loop from line 37 performs voting. So you can simply change 10 to 1 and set scale to  torch.ones(3, device=xyz.device). That should yield "Instance mIoU 87.28, Class mIoU 85.87". You can also disable (comment out) part_seg_refinement in line 48, which should lead to performance being the same as in training logs.
④ As the random seeds are not fixed, voted performance can fluctuate a bit, but should be near reported ones with high probability. Sometimes better, sometimes worse, but on average, near reported. It is also possible that your environment is different from mine that causes discrepancy. But I believe it's the first reason.

**Comment by zouyanmei** [comment]
> Hi. ① I mostly followed [PointNeXt#profile.py](https://github.com/guochengqian/PointNeXt/blob/master/examples/profile.py) for profiling. The basic logic is simple. Load a batch of inputs (I always load from dataloader), use deepspeed for FLOPs and params; do multiple forwarding to get throughput. ② Currently I'm occupied with other commitments. So at least for the time being, I have no plans for that (or even update the current version of code). ③ I assume you mean ShapeNetPart. The for loop from line 37 performs voting. So you can simply change 10 to 1 and set scale to torch.ones(3, device=xyz.device). That should yield "Instance mIoU 87.28, Class mIoU 85.87". You can also disable (comment out) part_seg_refinement in line 48, which should lead to performance being the same as in training logs. ④ As the random seeds are not fixed, voted performance can fluctuate a bit, but should be near reported ones with high probability. Sometimes better, sometimes worse, but on average, near reported. It is also possible that your environment is different from mine that causes discrepancy. But I believe it's the first reason.

Thanks for your kind reply. I have calculated on ScanObjectNN, but have problems on S3DIS, because it is hard to initial the value of indices. Can you give me some advice? Thanks again.

**Comment by binjiechen** [comment]
You're welcome. The indices are calculated in the dataset (s3dis.py). It's much easier to just load one as in the training script.
Here are simple steps you can follow to derive a profiling script from train.py.
① Copy the script, delete everything after line 74 (warmup_fn(xxx)).
② Before line 74, insert these lines:
```python
# a batch of 16x15000 points
# downsample ratio fixed to 4, as grid sampling has randomness
batch_size = 16
s3dis_warmup_args.max_pts = 15000
s3dis_warmup_args.grid_size = [0.04, 4, 4, 4]
```
③ Inside warmup_fn, replace model.train() with model.eval(). Then you can use the xyz, feature, indices below as model input for profiling.

Some explanation. During warmup, each point cloud in a batch is assured to have max_pts points. Letting the model process a possibly largest point cloud stabilizes memory consumption.

**Comment by zouyanmei** [comment]
> You're welcome. The indices are calculated in the dataset (s3dis.py). It's much easier to just load one as in the training script. Here are simple steps you can follow to derive a profiling script from train.py. ① Copy the script, delete everything after line 74 (warmup_fn(xxx)). ② Before line 74, insert these lines:
> 
> ```python
> # a batch of 16x15000 points
> # downsample ratio fixed to 4, as grid sampling has randomness
> batch_size = 16
> s3dis_warmup_args.max_pts = 15000
> s3dis_warmup_args.grid_size = [0.04, 4, 4, 4]
> ```
> 
> ③ Inside warmup_fn, replace model.train() with model.eval(). Then you can use the xyz, feature, indices below as model input for profiling.
> 
> Some explanation. During warmup, each point cloud in a batch is assured to have max_pts points. Letting the model process a possibly largest point cloud stabilizes memory consumption.

Thanks a lots. I have solved this problem, but the results on ShapeNetPart are different from the released. (7.6M and 2.95G, rather than 7.5M and 1.85G.) Shoud I report the reproduced results rather than the released?

**Comment by binjiechen** [comment]
I think you got it wrong somehow. I re-profiled it and I'm certain no mistake is made.
The 7.6 vs 7.5 is because I removed params for the regularization branch (cor_head).
As for FLOPs, see it this way. ShapeNetPart and ModelNet40 share a similar configuration.
ModelNet40: 96-192-384, 1024-256-64
ShapeNetPart: 96-192-320-512, 1024-256-96-32
As depth is the same for all stages. FLOPs for one stage can be estimated by dim * dim * points.
FLOPs for the first 3 stages are almost the same. Last block adds around 1/3. Seg head has little.
For the first stage, spatial encoding is more consuming. So I'd say 1.85G is reasonable given it's 1.44G on ModelNet40.

**Comment by binjiechen** [comment]
I know now. You set input points to 1024. But what we should measure is actually FLOPs per 1k points.

**Comment by zouyanmei** [comment]
> I know now. You set input points to 1024. But what we should measure is actually FLOPs per 1k points.

Yes, I think so. But Tab.1 in DeLA show that "FLOPs are counter per 1024 points."

**Comment by binjiechen** [comment]
Yes, per 1024 points, or 1k points. I mean 1024, not 1000. The point here is to measure FLOPs per n points, not when input points equals n.

**Comment by zouyanmei** [comment]
Thanks a lot. I am sorry that I cannot understand. I have another problem that should i report the results （7.6M and 3.7G with 64*2048 points）in my paper? The configuration is same as PointNeXt.

**Comment by binjiechen** [comment]
Yes, 3.7G is the FLOPs per 2048 points. It's a fair comparison as long as all methods use the same setting. For example, the network input is 2048 points and reported FLOPs is per 2048 points. For another example, on S3DIS, if all use an initial grid size of 0.04, and reported FLOPs is per 15000 or 1024 points, it is fair then.

**Comment by zouyanmei** [comment]
> Yes, 3.7G is the FLOPs per 2048 points. It's a fair comparison as long as all methods use the same setting. For example, the network input is 2048 points and reported FLOPs is per 2048 points. For another example, on S3DIS, if all use an initial grid size of 0.04, and reported FLOPs is per 15000 or 1024 points, it is fair then.

I got it, thanks again.
